### ISSUE TITLE
Placing multiple orders

### ISSUE THREAD

**Issue by gspasic** [JKorf/Binance.Net#355]
When trying to create async or synchronous orders in several threads at once the orders are returning null and not being placed. I rarely have this issue when placing orders in series, one after the other. It can take up to 10 seconds to place all the orders I need and I really want the process to be a lot quicker. Is there any way to speed up the process of creating/cancelling multiple orders?

**Comment by thomasd3** [comment]
Jan just merged the API call to place multiple orders in the main branch; Binance allows up to 5 orders / cancellations per call though, not more. I've been using that code for a few days and orders get placed quickly that way.

**Comment by gspasic** [comment]
Thanks! Do you know if this works with spot orders or is it just for futures?

**Comment by thomasd3** [comment]
it's in the futures' API

**Comment by gspasic** [comment]
Oh ok. I'm only trading spot orders.

**Comment by JKorf** [comment]
This would be an issue on Binance side. The library is thread safe sending the requests to Binance. If Binance returns null there should be an error in the Error property in the result. You could have a look to see what the error is that is returned.

**Comment by gspasic** [comment]
> This would be an issue on Binance side. The library is thread safe sending the requests to Binance. If Binance returns null there should be an error in the Error property in the result. You could have a look to see what the error is that is returned.

Simple, yet great advice! I actually wasn't aware of the error codes. I'll look into it.

**Comment by gspasic** [comment]
Ok, so I've been running orders over night and the most common error that I am getting is "1022: Signature for this request is not valid". Other times it would fail to get an order with "2013: Order does not exist", then try again in about 30 seconds and be successful.
I'm totally guessing that this may be a timestamp issue since this happens intermittently and the error goes away without reconnecting to the binance API? 
@JKorf if you had any insight to what is going on I would really appreciate it! Thanks again for the help!

**Comment by developocalypse** [comment]
Automatically syncing your system clock with a time server every five minutes should fix intermittent invalid signature errors.

**Comment by thomasd3** [comment]
I'm running it on several AWS servers (I assume their time is kept in sync) and we're doing market making, so there are a lot of orders placed and canceled all the time. I never had that error, so I believe it's not in the code itself, but, as @developocalypse mentioned, it's a server time issue.

**Comment by JKorf** [comment]
As suggested, please try to sync your system clock. Binance is very strict in that regard. 

I'm not aware of any restrictions on Binance side regarding the order of received timestamps (say a later timestamp is received before a request send earlier), but that might be something we can look into if the system clock syncing doesn't help.

**Comment by gspasic** [comment]
So I've tried manually syncing the time via windows adjust date/time and I've also created a script to do it through command prompt every 5 minutes or so. Unfortunately it hasn't worked. I am still getting the errors at about the same frequency as before.

**Comment by developocalypse** [comment]
If there's an issue with your script you could try this:
www.thinkman.com/dimension4
The user interface is non-standard; if you choose this and cannot get it syncing correctly every five mins and loading at startup, post here and I'll help.

**Comment by gspasic** [comment]
Thanks @developocalypse I've got it working - looks like a handy little app! It said my time was out by about 300ms on the first sync so hopefully this does the job! I'll give it a day or so and let you know how it goes.

**Comment by gspasic** [comment]
Ok, just to give you an update. I tried running my with my system time being updated every minute or so but it had no effect. I've switched back to running my methods which call the PlaceOrderAsync etc. methods in a single thread and I am no longer getting the signature error. I was calling my methods with the below code to spin up a thread for each coin - I guess this isn't the best way to do it since it caused me so many issues:

for (int num = 0; num < coins.Count; num++)
{
     int loopnum = num;
     Task.Factory.StartNew(() => { actionSignals(loopnum); });
}
